### ISSUE TITLE
typo

### ISSUE THREAD

**Issue by yarimitsu** [quantifish/isotope#18]
I found a little typo on lines 58-60 in sigma_N[g] that was responsible for making those sigma_phi.3s so weird. I fixed it and things are much better now. 

It got me wondering why there are 3*G sigma_phis. I know we've gone back and forth about the name a couple times, and I am inclined to go with whatever you think because for the most part I'm just trying to keep up. But this group of parameters represents variability in the sources (i.e., dC[3], dH[3], and dN[3]) not phis. What we call it doesn't matter. However, I believe we should be using a 3x3 matrix for sigma_phi rather than a 3xG matrix, because sources are shared among species.

It's not changed yet, but if there's no reason not to change it I'll deal with it tomorrow.

**Comment by quantifish** [comment]
Good work on typo.

Well one of us is confused here. I thought that the sigma_phi parameters was supposed to represent variability in the phis and that the sigma_frc parameters were the variability in the sources. I just checked the code and see that at some point this all got changed so that we do have way too many sigma parameters.

The problem here is the equations that we are supposed to be using, and your document is a little bit lacking on details. Once we know the equations the code is super easy to write. Unfortunately you seem to be trying to learn the math from a bunch of old dudes that cant do math themselves. I have helped out in the word docco that you started on. This has some comments in it from me, and needs work!

**Comment by quantifish** [comment]
Give me a few minutes and Ill push

**Comment by quantifish** [comment]
OK, so more work needed on the docco/math before we can go any further with the plots too I think. Need to see what that dH_adj bit is doing in that document too. The document needs to explain all of the code in eqn form.

**Comment by yarimitsu** [comment]
Well it's good that someone knows something about math here, because it's
obvious that I don't.

Fractionation is anything that may cause change in isotope ratios
(metabolic processes related to trophic transfer and other things) so
related to Delta_C/C_tot and Delta_N/N_tot, and Delta_H for sure. Probably
also needs to be sigma_omega in there also, but we're treading new water on
this for the most part.

 I'll try to fix up the docco now. Thanks for the start, and for suffering
through this with me. This project needles at my weaknesses in ways I never
imagined...

On Aug 25, 2015, at 10:44 PM, Darcy Webber <notifications@github.com> wrote:

Good work on typo.

Well one of us is confused here. I thought that the sigma_phi parameters
was supposed to represent variability in the phis and that the sigma_frc
parameters were the variability in the sources. I just checked the code and
see that at some point this all got changed so that we do have way too many
sigma parameters.

The problem here is the equations that we are supposed to be using, and
your document is a little bit lacking on details. Once we know the
equations the code is super easy to write. Unfortunately you seem to be
trying to learn the math from a bunch of old dudes that cant do math
themselves. I have helped out in the word docco that you started on. This
has some comments in it from me, and needs work!

—
Reply to this email directly or view it on GitHub
<https://github.com/quantifish/isotope/issues/18#issuecomment-134862205>.

**Comment by yarimitsu** [comment]
Don't know if it's going to help or hurt, but I made some changes to the model documentation, with some comments about a bit of confusion that i need to work out. 

Consumer isotope data only plot nicely with source data if you deal with fractionation. People typically present data relative to sources (ie, the isotope mixing space) in one of two ways: either adjust the consumer mixture data down to sources (dC_exp - C_tot), or adjust the source values up to the mixture data (dC + C_tot). This is a so-called fractionation correction, and has to be done in order to visualize the isotope mixing space. Hopkins et al 2012 argues that it's better to adjust the sources up rather than the mixtures down. That's essentially what I did with the srcC, srcH, and srcN generated quantities.

For Hydrogen there's another issue: the consumer tissue data (dH_obs) must be adjusted for the proportion of d2H that is due to diet (i.e., 1-omega) because that's where the (fractionation corrected) source data are relevant. I used the mean of the posterior distribution for each dH_adj_i for this...but maybe that's not the right thing to do. 

I think there is value to the isospace plot, but to my knowledge it hasn't been done for hydrogen before. This adjustment might need to be rethought. 

I see some other places in the model documentation that I'm confused about, partly notation, but partly the nuts and bolts of what the model is doing. E.g...why mean(omega_g * dH_w_g). That seems like something dumb I did (a hack!) to make the code run. There is a lesson here regarding reactive vs proactive programming...I'm afraid I have some terrible habits.

I have to run to the airport to pick up Chris, might not get back to thinking about this for a few hours.

**Comment by quantifish** [comment]
I'm going to recommend we don't do any more modelling or plotting until 
the document and the equations are sorted, all the confusion is making 
us go round in circles. To be honest, the notation in the document is 
very confusing, and all the papers are much the same.

I agree that the use of mean() throughout the model is likely a hack, 
again going through the equations is going to help here.

I will take a look at the docco and make comments.



On 15-08-27 10:08 AM, yarimitsu wrote:
>
> Don't know if it's going to help or hurt, but I made some changes to 
> the model documentation, with some comments about a bit of confusion 
> that i need to work out.
>
> Consumer isotope data only plot nicely with source data if you deal 
> with fractionation. People typically present data relative to sources 
> (ie, the isotope mixing space) in one of two ways: either adjust the 
> consumer mixture data down to sources (dC_exp - C_tot), or adjust the 
> source values up to the mixture data (dC + C_tot). This is a so-called 
> fractionation correction, and has to be done in order to visualize the 
> isotope mixing space. Hopkins et al 2012 argues that it's better to 
> adjust the sources up rather than the mixtures down. That's 
> essentially what I did with the srcC, srcH, and srcN generated quantities.
>
> For Hydrogen there's another issue: the consumer tissue data (dH_obs) 
> must be adjusted for the proportion of d2H that is due to diet (i.e., 
> 1-omega) because that's where the (fractionation corrected) source 
> data are relevant. I used the mean of the posterior distribution for 
> each dH_adj_i for this...but maybe that's not the right thing to do.
>
> I think there is value to the isospace plot, but to my knowledge it 
> hasn't been done for hydrogen before. This adjustment might need to be 
> rethought.
>
> I see some other places in the model documentation that I'm confused 
> about, partly notation, but partly the nuts and bolts of what the 
> model is doing. E.g...why mean(omega_g * dH_w_g). That seems like 
> something dumb I did (a hack!) to make the code run. There is a lesson 
> here regarding reactive vs proactive programming...I'm afraid I have 
> some terrible habits.
>
> I have to run to the airport to pick up Chris, might not get back to 
> thinking about this for a few hours.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/quantifish/isotope/issues/18#issuecomment-135190420>.
>

-- 
--------------------------------------------------
D'Arcy Webber
Quantifish | Quantitative Marine Science Solutions
P: +6421 0233 0163
E: darcy@quantifish.co.nz
W: www.quantifish.co.nz
--------------------------------------------------

**Comment by quantifish** [comment]
OK, added comments to the docco. Needs work. Also, please use track changes as I needed to redo some stuff.

I redid the primary eqns. They are much easier to follow now I think, take a look, and repeat for C (write these out explicitly, it makes the notation much cleaner).

**Comment by quantifish** [comment]
You getting through it Yumi? Making sense?

**Comment by yarimitsu** [comment]
Thanks for checking in Darcy! YES, this is so good. I wish I would have
done this ages ago...but I had no idea how. Thanks for setting me straight
on that notation, things are making so much more sense. That math you wrote
DOES read easy as pie now.

I'm so almost done (waylaid by things like the POS Word crashing and not
recovering my track changes, and adminy government work, and my niece who
just called for some life advice, and I had to check in with an employee
I've been ignoring for 2 weeks and... and blah blah). Will send it up in a
few minutes...

**Comment by quantifish** [comment]
No worries, happy to take a look at the next version. Once we are both 
happy with the document, the notation and the math, we can go back to 
the code so that it reflects the math, then come back to the math and 
think of any improvements that we can make, then go back to the code, 
back to the plots, and so the endless circle goes. But, I do think it is 
getting close!

Just while I am yarning, you'll be intending to publish this right? Any 
thoughts on what journal you might aim for? I think you could get this 
work into "Ecological Modelling", which may not be the first journal 
that pops into your mind, but it always looks good to have some 
modelling journal publications up your sleeve!

Haha, sounds very much like my life. I'm trying to work two jobs, 
publish two papers, start writing a book, and pick up additional 
contracts for next year. Which reminds me I should email Chris about the 
tag-recapture thing...


On 15-08-28 09:25 AM, yarimitsu wrote:
> Thanks for checking in Darcy! YES, this is so good. I wish I would have
> done this ages ago...but I had no idea how. Thanks for setting me straight
> on that notation, things are making so much more sense. That math you 
> wrote
> DOES read easy as pie now.
>
> I'm so almost done (waylaid by things like the POS Word crashing and not
> recovering my track changes, and adminy government work, and my niece who
> just called for some life advice, and I had to check in with an employee
> I've been ignoring for 2 weeks and... and blah blah). Will send it up in a
> few minutes...
>
>
>
> On Thu, Aug 27, 2015 at 1:06 PM, Darcy Webber <notifications@github.com>
> wrote:
>
> > You getting through it Yumi? Making sense?
> >
> > —
> > Reply to this email directly or view it on GitHub
> > 
> <https://github.com/quantifish/isotope/issues/18#issuecomment-135554980>.
> >
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/quantifish/isotope/issues/18#issuecomment-135558688>.
>

-- 
--------------------------------------------------
D'Arcy Webber
Quantifish | Quantitative Marine Science Solutions
P: +6421 0233 0163
E: darcy@quantifish.co.nz
W: www.quantifish.co.nz
--------------------------------------------------

**Comment by quantifish** [comment]
Just added some more to docco

**Comment by yarimitsu** [comment]
I think MS Word contends with that other thing for biggest POS ever written. I just crashed that docco 5 times using 2 MSWord versions, and changes were unrecoverable most of the time. So along with other junk I've been doing, that's why it took so dang long!

Ecological Modeling could be a  good place for this. I've been sort of waiting for the results to come in before deciding how big to go :). But yes, this is definitely going to be published sooner than later, with you on it obv. 

BUT, I feel like a new and (hopefully) slightly less stupid person having your help on the notation. Not perfect yet but a whole new understanding of things for sure. There are definitely some places to change the code after sorting this out.

You're fast. OK I'm back on it. Priors and stuff...

**Comment by yarimitsu** [comment]
It's impossible to keep up with you.

**Comment by quantifish** [comment]
Yeah word is junk. Latex is a lot better but a steep learning curve 
unfortunately.

Yes, a fair few changes to the model required now that things are 
becoming clear! I was gonna take a 10 minute look at the model now, I 
won't touch the docco for now.


On 15-08-28 12:57 PM, yarimitsu wrote:
>
> I think MS Word contends with that other thing for biggest POS ever 
> written. I just crashed that docco 5 times using 2 MSWord versions, 
> and changes were unrecoverable most of the time. So along with other 
> junk I've been doing, that's why it took so dang long!
>
> Ecological Modeling could be a good place for this. I've been sort of 
> waiting for the results to come in before deciding how big to go :). 
> But yes, this is definitely going to be published sooner than later, 
> with you on it obv.
>
> BUT, I feel like a new and (hopefully) slightly less stupid person 
> having your help on the notation. Not perfect yet but a whole new 
> understanding of things for sure. There are definitely some places to 
> change the code after sorting this out.
>
> You're fast. OK I'm back on it. Priors and stuff...
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/quantifish/isotope/issues/18#issuecomment-135594364>.
>

-- 
--------------------------------------------------
D'Arcy Webber
Quantifish | Quantitative Marine Science Solutions
P: +6421 0233 0163
E: darcy@quantifish.co.nz
W: www.quantifish.co.nz
--------------------------------------------------

**Comment by quantifish** [comment]
Haha well that's because you're doing all the hard work.


On 15-08-28 12:58 PM, yarimitsu wrote:
> It's impossible to keep up with you.
>
> On Thu, Aug 27, 2015 at 4:49 PM, Darcy Webber <notifications@github.com>
> wrote:
>
> > Just added some more to docco
> >
> > —
> > Reply to this email directly or view it on GitHub
> > 
> <https://github.com/quantifish/isotope/issues/18#issuecomment-135593658>.
> >
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/quantifish/isotope/issues/18#issuecomment-135594737>.
>

-- 
--------------------------------------------------
D'Arcy Webber
Quantifish | Quantitative Marine Science Solutions
P: +6421 0233 0163
E: darcy@quantifish.co.nz
W: www.quantifish.co.nz
--------------------------------------------------

**Comment by yarimitsu** [comment]
OK I'm working on the docco now! won't touch the model

**Comment by yarimitsu** [comment]
HA. that's not true.

**Comment by yarimitsu** [comment]
I should remove the parentheses and ^2 on these, so I'm talking about SDs rather than var, right?

![image](https://cloud.githubusercontent.com/assets/13755255/9536980/344ea5c4-4cdf-11e5-8304-9f56db092213.png)

**Comment by quantifish** [comment]
Yes, correct. While I am at it, in the doc you have a sigma_frc for each species, in the code you have specified a single sigma_frc for all species. Which is it? Is the doc correct? Also, in the doc you have put in a sigma_omega for each species. Is this also correct?

**Comment by quantifish** [comment]
Eqn numbers would be useful so we can reference eqns

**Comment by yarimitsu** [comment]
Ok i'll put eq #s in.

sigma_frc[2,G] is correct - that's one for each species for carbon and
nitrogen. These are about C_tot and N_tot.

sigma_frcH is shared across all species, so just one of those for
everything. THis is about Delta_H (which is the only level we think has
true fractionation)

**Comment by yarimitsu** [comment]
Also, sigma_omega[g] needs to be added as a parameter to the code. I assume
that it's necessary...as a source of variability in H. It wasn't in there
before because the model I was working (the Semmens thing, with u_hats)
didn't include hydrogen...

**Comment by yarimitsu** [comment]
Don't look at isotope2.doc yet, I just put it up there so I could get your changes to the model.

**Comment by yarimitsu** [comment]
we've changed sigma_phis [3, G] to this in the model:   
real<lower=0> sigma_src[3];      // sigma source parameters dC, dH, dN  

but because of these equations, where k = 3:
![image](https://cloud.githubusercontent.com/assets/13755255/9555387/2b6de95c-4d79-11e5-88b4-827b7e762f24.png)

![image](https://cloud.githubusercontent.com/assets/13755255/9555338/d70c2e6e-4d78-11e5-96e8-be0676ca9da6.png)

I think it should be this:
real<lower=0> sigma_src[3, 3];      // sigma source marine, coastal and FW parameters for dC, dH, dN

Any objections?

**Comment by quantifish** [comment]
Correct you are, good stuff Yumi! Change away!


On 15-08-29 07:40 AM, yarimitsu wrote:
>
> we've changed sigma_phis [3, G] to this in the model:
>
> real sigma_src[3]; // sigma source parameters dC, dH, dN
>
> but because of these equations, where k = 3:
> image 
> <https://cloud.githubusercontent.com/assets/13755255/9555387/2b6de95c-4d79-11e5-88b4-827b7e762f24.png>
>
> image 
> <https://cloud.githubusercontent.com/assets/13755255/9555338/d70c2e6e-4d78-11e5-96e8-be0676ca9da6.png>
>
> I think it should be this:
> real sigma_src[3, 3]; // sigma source marine, coastal and FW 
> parameters for dC, dH, dN
>
> Any objections?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/quantifish/isotope/issues/18#issuecomment-135869478>.
>

-- 
--------------------------------------------------
D'Arcy Webber
Quantifish | Quantitative Marine Science Solutions
P: +6421 0233 0163
E: darcy@quantifish.co.nz
W: www.quantifish.co.nz
--------------------------------------------------

**Comment by yarimitsu** [comment]
OK here's another place where there's discrepancy and a spot where my math ignorance runs deep. How do I fix this? Which (if any) is correct math?
The model:
 for(gg in 1:G) {
      srcC[1,gg] <- dC[1] + C_tot[gg] + sigma_C[gg];
      srcC[2,gg] <- dC[2] + C_tot[gg] + sigma_C[gg];
      srcC[3,gg] <- dC[3] + C_tot[gg] + sigma_C[gg];
      srcH[1,gg] <- dH[1] + Delta_H   + sigma_H[gg];
      srcH[2,gg] <- dH[2] + Delta_H   + sigma_H[gg];
      srcH[3,gg] <- dH[3] + Delta_H   + sigma_H[gg];
      srcN[1,gg] <- dN[1] + N_tot[gg] + sigma_N[gg];
      srcN[2,gg] <- dN[2] + N_tot[gg] + sigma_N[gg];
      srcN[3,gg] <- dN[3] + N_tot[gg] + sigma_N[gg];
    }  

The docco:
![image](https://cloud.githubusercontent.com/assets/13755255/9556542/0030f10e-4d82-11e5-8050-6ba140d35f8a.png)

**Comment by quantifish** [comment]
OK, so forget about the code, lets get the math right then worry about that.

I'm a little confused with which docco to follow, better to stick with a single docco and use track changes or place notes/draft/old stuff down the bottom if you want.

Basically, the equations above are confusing and I don't follow what they are doing or why. Please change the fractionation corrected sources from sf^C_jk to something a little cleaner. How about the gamma symbol, so we would have 

gamma^C_jk

Now , the next bit suggests we have another likelihood going on in this model which is confusing. Can you take a look at the doccos again, merge into a single docco, then revisit this part of the docco and make sure we actually need it? If you think we do, pop a reference in there so I can take a look at where this is coming from...

**Comment by quantifish** [comment]
P.S. Ive added notes comments and eqns to isotope.doc

**Comment by yarimitsu** [comment]
I made changes to the docco. Not all changes are tracked because the comments + track changes + equations cause the software to freak out and can't save the doc like that. I made all your suggested changes, I left other track changes in there wherever possible...

I noticed the old Hopkins et al 2012 paper was a misnamed pdf pointing some old paper, I fixed it so now the right paper is in there. Also added some other relevant papers that I referenced in the docco.

Yesterday I messed with the plots (some comments in there) and a little with the model code (ie sigma_src from 3*G to 3*3). Need your advice on how to deal with uncertainty in the fractionation-corrected sources, which are generated quantities that JUST for plotting purposes (so this shouldn't require another likelihood eq?)...I took a stab at it in the docco (could be wrong still) but nothing is changed there in the code.

**Comment by quantifish** [comment]
OK, back at ya with the docco. Also, somewhere along the way the model got stuffed up by GitHub, this happens when we both work on the same thing at the same time then try and write over top of each other. I'll leave it to you to fix this up and get it working again. Once the model is a go again I can take a look at it and the plots.

**Comment by yarimitsu** [comment]
OK I see what I did to mess that up last time. Sorry about that.

I've got the writing from here if you think the equations are the way they
should be. Thanks for the help with the docco.

**Comment by quantifish** [comment]
Hey Yumi. I've broken the code a little, will fix tomo when I get up, Stuff to do on the docco anyway.
